### PR TITLE
Correction of the csum problem, and addition of a numerical stability…

### PR DESCRIPTION
**Pull request by ebeauce**
In the wrapper function, it has to be specified to the numpy.csum function that we want float64 numbers. Otherwise, the csum is computed in float32, and then casted in float64 when we call np.float64(csum). I made the test, it makes a BIG difference. Still, I get CC that are slightly larger than 1. (up to 1.3 in my real data test), so the problem is not over. However, I doubt we can overcome this problem without changing everything to double-precision floating-point numbers (or going back to single-precision for the csum).

I believe the problem associated with NaNs is originating from something else. When the denominator is too close from zero, or zero, it clearly raises an undefined behavior in the division. Even if we could expect to have the CCs messed up only at the location where this problem occurs, it seems the whole time series is affected by those NaNs.

Just adding a condition that check whether the denominator is not too low solves this problem. Therefore, I added such a safety condition in the two versions.

Calum, if you can check your NaN problem is solved with the last version of matched_filter.c and fast_matched_filter.py, it would be great ! But still you should encounter a problem of precision (one problem at a time !).

### CODE REVIEW

**Comment by calum-chamberlain** [comment]
@ebeauce I think you should merge PR #8 into master, then update this branch (just use the update button), then we can have a nice clean git history and run the tests without having to muck around moving files about and creating a messy repository.

**Comment by ebeauce** [comment]
William and Eric here: we're trying to figure out if we have a bug or not. In the test case you set up, is your template supposed to find itself or a repeat within the day of test data?  We ask, because we don't find anything.

Analogous to the synthetic test, we generated three fake templates of noise from the Kaikoura data to test our functions.  We took one noise window before the EQ (quiet), one noise window during the EQ (noisy), and one noise window after the EQ (to check if the mean has changed after the EQ).  All three of the fake noise templates work and find themselves with no difficulty.

However, with your template we don't find anything and we just have a persistent correlation coefficient noise.  Is this intended?

**Comment by calum-chamberlain** [comment]
Hey chaps, briefly from the field: no, the template is not included in the continuous data, the template is from a few hours earlier, I ended up timing the data around the mainshock which was where the nan errors were coming from. If nans are gone out should be good. The reason for trimming the data was to reduce the file size to make sure the repo wasn't toooo big. Sorry for not telling you this. 

From memory there should be done detectable events at 8*MAD in the continuous data attached, but there is definitely no perfect match.

Good work guys, from the commit messages it looks like you have gone with a similar approach to the GPU summing? Sounds like that should be the most stable. 

It might be a good idea to add a test that, when both GPU and CPU routines are available, compares the two to make sure they generate comparable results? I can add one of you want?

**Comment by calum-chamberlain** [comment]
So I wrote a comparison test, which will skip the test if either the CPU or GPU code is not loaded - I put this in a branch off `eric` and have opened a PR here: #10

It looks like the GPU routines work well, and no NaN's appear now, but the noisy dataset is not well bounded for the CPU case.  I'm not sure why this is, it seems well bounded for the GPU case.

**Comment by beridel** [comment]
HI Calum.  We now know that the accuracy problem for the CPU code is the cumulative sum to compute the sum of squares of the data.  See here for a description of the issue: https://en.wikipedia.org/wiki/Kahan_summation_algorithm

We're trying on getting that implemented.  Once that's good, we can back to everything being a float and we should have the accuracy we need.

File: fast_matched_filter/src/matched_filter.c
Comment (calum-chamberlain) [inline fast_matched_filter/src/matched_filter.c]: I think this will result in datasets with small variances not being normalised.  I have been running into this issue [here](https://github.com/eqcorrscan/EQcorrscan/pull/182), but we use a double for the normalisation then cast back to float.  It doesn't seem to add much memory overhead (you are only storing one double in memory, and casting to float), but it may increase the time.  It might be worth a test because it will allow you to apply these codes to a greater range of data.
I also think it is worth making the user aware that their variance is too small and that their correlations are not being normalised when the denominator is < 0.00001.
This is a rare case for passive seismic data, so up to you, but worth making sure that any users know the limitations.

File: fast_matched_filter/src/matched_filter.cu
Comment (calum-chamberlain) [inline fast_matched_filter/src/matched_filter.cu]: Same comment as the C code.

**Review by calum-chamberlain — CHANGES_REQUESTED** [review]
This still does not pass the test for the noisy dataset - the minimum correlation is less than -1.0 (I get -1.019), which is not far off, but hints that something isn't working properly.  Perhaps the issue in the CPU code that resulted in large correlation ranges (up to 15), that has been nullified by not normalising for small variances is rearing its head as another issue: if the variance is too small the correlations are not normalised resulting in exceeding the +/-1 range. It is hard to know for certain without knowing when the normalisation is not occurring.

**Comment by ebeauce** [comment]
Calum,

Regarding your comments on the addition of "if (denominator > 0.00001)", I don't agree with you saying that if the denominator is smaller than 0.00001, the CC won't be normalized. If the denominator is very small, the computation of the CC is just skipped so it results in CC=0. From what I've experienced, this helps a lot when you have gaps in the data that were filled with 0's.

Also, since the modification of CPU_LOADED and GPU_LOADED, I get an import error when I try to run your test script. As the beginning, it tries to import:
from fast_matched_filter import (matched_filter, CPU_LOADED, GPU_LOADED)
and triggers:
ImportError: cannot import name CPU_LOADED
Is it working well for you ?

Also, in your test you generate the weights with:
weights = np.ones(
                (len(dataset['templates']), len(dataset['data'])))
and they are not normalized so that they sum to one. This means you are not expecting the CCs to be bounded between -1 and +1, but between +/-sum(weights) (i.e. +/-6 in your example). If I understand the import error, I'll be able to check if the problem comes from the normalization of the weights.

**Comment by calum-chamberlain** [comment]
That seems fair on the skipping computation of cc and setting to zero - that can't be the thing that does not generate bounded correlations (more on that below). This will impact on thresholding if using some threshold based on the distribution of the cross-correlations (e.g. MAD), for example, if you have multiple zeros in your correlation vector your MAD will be zero. I do think it is worth warning the user when this is happening.

Yes these imports work fine - have you reinstalled fast_matched_filter? You may need to manually remove the old version (setup.py and pip sometimes install with different names making it difficult to keep track of what is what).

Interesting, I had assumed that the code would normalise, I can change the weights, but when I change them to be `1/len(dataset['data'])` I get a maximum cross-correlation at 2.5... Am I going the wrong way?

When I run the same dataset through EQcorrscan's correlation routines I get a maximum cross-correlation sum (where I would expect it to be bounded by +/-5) of 0.399.

**Comment by calum-chamberlain** [comment]
That seems fair on the skipping computation of cc and setting to zero - that can't be the thing that does not generate bounded correlations (more on that below). This will impact on thresholding if using some threshold based on the distribution of the cross-correlations (e.g. MAD), for example, if you have multiple zeros in your correlation vector your MAD will be zero. I do think it is worth warning the user when this is happening.

Yes these imports work fine - have you reinstalled fast_matched_filter? You may need to manually remove the old version (setup.py and pip sometimes install with different names making it difficult to keep track of what is what).

Interesting, I had assumed that the code would normalise, I can change the weights, but when I change them to be `1/len(dataset['data'])` I get a maximum cross-correlation at 2.5... Am I going the wrong way?

When I run the same dataset through EQcorrscan's correlation routines I get a maximum cross-correlation sum (where I would expect it to be bounded by +/-5) of 0.399.

**Comment by ebeauce** [comment]
I think the case where you would get a MAD equal to zero is really extreme, because it would suppose that the CC computation is skipped on every component and station during the whole day (or the whole period that the user is processing at the same time). I can only imagine two cases where that could happen:
- the data is equal to zero everywhere,
- the template is zero everywhere.
Which is clearly not an interesting situation. Do you agree with that ?

To me, the computation is only skipped when the CC should truly be 0, but because of limitations in the numerical precision, we end up with undefined behaviors (anyway, 0/0 is not defined for a computer).

For the normalization of the weights, I think we should emphasize in the docstring if they are normalized or not. Multiplying the array by 1/len(dataset['data']) should work. I just add this line in the script:
```python
weights /= len(dataset['data'])
```
and the tests are passed correctly. Here is a copy of the output I get:

```bash
Computing correlations for dataset noisy
Trying to compute using the cpu
Trying to compute using the gpu

Computing correlations for dataset kaik
Trying to compute using the cpu
Trying to compute using the gpu
NaN checking for: noisy_gpu
NaN checking for: kaik_cpu
NaN checking for: noisy_cpu
NaN checking for: kaik_gpu
.Checking reasonable boundedness for correlations for noisy_gpu
Max of data is: 0.155925
Min of data is: -0.169886
Checking reasonable boundedness for correlations for kaik_cpu
Max of data is: 0.079792
Min of data is: -0.074930
Checking reasonable boundedness for correlations for noisy_cpu
Max of data is: 0.155925
Min of data is: -0.169886
Checking reasonable boundedness for correlations for kaik_gpu
Max of data is: 0.079792
Min of data is: -0.074930
.
----------------------------------------------------------------------
Ran 2 tests in 13.292s
```

**Comment by calum-chamberlain** [comment]
Hey Eric, agree on the weights test - good to see it pass, strange, but not unexpected that it gives different maxima to normalised correlations.  I tried to elude to my lack of confidence about how I was applying weights in [this comment](https://github.com/beridel/fast_matched_filter/pull/8#issuecomment-336617786), so I'm glad we got to the bottom of that.

~~I'm running into some other issues with the Kaikoura dataset that I am working on otherwise as well and I will share a jupyter-notebook with you soon so that you can get to the bottom of that.~~ Scrap that, issue stemmed from running an old version of fast_matched_filter. Re-running with this branch looks good! 🎉 🎈 

Regarding the zero MAD cases.  I was in the same mindset as you until the start of this week, when a user got in touch about some active-source data he is playing with.  In this case the SNR is incredibly high, with background amplitudes around 10e-7. Running this through EQcorrscan's routines with a similar minimum threshold to you resulted in background correlations being zeroed (even though there were valid correlations there).  Because there were only a few sources in the trace being examined, there were only a few non-zero correlations.  When computing the median of this (depending on precision here as well) we ended up with a zero median. That means that anything that has an amplitude above the threshold we set to be able to compute correlations was classified as being above the MAD and therefore a detection.  This is strange behaviour, and would be particularly vexing if you were trying to discriminate between different sources using correlation.

I do agree that this is an edge-case, EQcorrscan has been out and about for a few years now and we have played with active-source data ourselves previously and not found this issue, so I wouldn't sweat it, but, because we have found that it can happen, I think you can learn from our failures and at-least warn the user when this case occurs so that they understand what is happening.  The worst thing we could do with our respective codes is give a user false hope that they have a robust result without understanding the limitations.

The changes that I made to reduce the impact and warn the user are [here](https://github.com/eqcorrscan/EQcorrscan/pull/182/files), particularly the `multi_corr.c` and `correlate.py` files. You are welcome to take what you want. Essentially to warn, we include an int variable that counts the number of correlations missed.  If this is more than 0 at the end then the return value is set to a known error code, which the Python function checks and raises a warning.

Does that make sense? I would share the data that raised this issue for us, but it's not my data to share unfortunately. If you want I can ask if I can share it though?

**Comment by ebeauce** [comment]
I'm glad the test is now running well !

I agree that we shouldn't neglect this kind of edge-cases. I didn't know that such lucky people dealing with ambient noise that low were actually existing ! I'm going to have a look at what you do to warn the user, and I'll think of something. I think the user should be warned when many CCs are turned to 0, and have an easy to tune the threshold as he wishes (like adding a "#DEFINE threshold" at the beginning of the scripts).

For the data, I can also generate synthetic data with very low std to check the behavior of FMF with very low values.

**Comment by beridel** [comment]
Are we good to merge this? I'm not sure what to do with the requested changes you made Calum (not what to actually do about it, but how to make sure the pull request is "clean")...

**Comment by calum-chamberlain** [comment]
It sounded like @ebeauce was going to make a change to add in a way to warn if correlations were not being computed? I will approve the changes anyway and you guys can do as you see fit.

**Review by calum-chamberlain — APPROVED** [review]

**Comment by beridel** [comment]
Alright, I'll leave it to Eric to add in such a warning and I will merge once that happens.  Sorry, I've not been following very closely what's been happening with the New Zealand trip.  Now that I'm here, I'll be better at following what's happening!

**Comment by ebeauce** [comment]
I messed up the commits again but it should be OK ! I added a warning message if more than 10 CCs are strictly equal to zero. I also added a #define STABILITY_THRESHOLD (the number to which the denominator is compared) at the beginning of the two scripts matched_filter.c/.cu just to give the option of tuning this parameter in a simple way.

**Review by calum-chamberlain — COMMENTED** [review]
Nice, good work, I probably would have put the check in the C-code, but should work in the wrappers - do you need to do something similar in the Matlab wrapper?

**Comment by ebeauce** [comment]
I'm really not a Matlab guy, so I tried to make some modifications to include this warning and also all the modifications on the cumulative sum computation, but it is just a draft for now.
I adapted the Matlab wrapper so that the cumulative sum is computed first, calling the C function csum, and then the main function matched_filter is called.
However, because I don't understand well how Matlab interacts with the C code, I have just added a new function in the _mex.c file by mimicking what William did before. This new function is supposed to call the C function csum. Obviously what I did doesn't work, but if someone could help me to put everything in order, it would be great !

**Comment by beridel** [comment]
I just updated the matlab code.  Until we have a good cumulative sum functionality, I've made it so the mex wrapper code calls sum directly, instead of calling it from matlab code.

I'm going to merge now because I think everything is working.

**Pull request merged**
